pkg/plugin/sdk/destination.go

```diff
@@ -0,0 +1,249 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sdk
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/conduitio/conduit-plugin/cpluginv1"
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+)
+
+// Destination TODO
+// All implementations must embed UnimplementedDestination for forward
+// compatibility.
+type Destination interface {
+	// Configure is the first function to be called in a plugin. It provides the
+	// plugin with the configuration that needs to be validated and stored. In
+	// case the configuration is not valid it should return an error.
+	Configure(context.Context, map[string]string) error
+
+	// Open is called after Configure to signal the plugin it can prepare to
+	// start writing records. If needed, the plugin should open connections in
+	// this function.
+	Open(context.Context) error
+
+	// WriteAsync receives a Record and can cache it internally and write it at
+	// a later time. Once the record is successfully written to the destination
+	// the plugin must call the provided AckFunc function with a nil error. If
+	// the plugin failed to write the record to the destination it must call the
+	// supplied AckFunc with a non-nil error. If any AckFunc is left uncalled
+	// the connector will not be able to exit gracefully. The WriteAsync
+	// function should only return an error in case a critical error happened,
+	// it is expected that all future calls to WriteAsync will fail with the
+	// same error.
+	// If the plugin caches records before writing them to the destination it
+	// needs to store the AckFunc as well and call it once the record is
+	// written.
+	// AckFunc will panic if it's called more than once.
+	// If WriteAsync returns ErrUnimplemented the SDK will fall back and call
+	// Write instead.
+	WriteAsync(context.Context, Record, AckFunc) error
```

Comment:
I introduced `WriteAsync` and `Write` to make it easier to write connectors that don't want to cache records but instead directly write them to the destination (the comments above the functions explain the behavior of each). I'm wondering if this interface makes sense or if it's just confusing? On the other hand, using `WriteAsync` can be a bit error prone if the developer forgets to call `AckFunc` because we expect to get acks for all handled records, regardless if an error happened or not. Acceptance tests would help here, but can't completely prevent all edge cases.

Feedback on this part is especially appreciated.